pkg/kubectl/cmd/edit.go

```diff
@@ -0,0 +1,419 @@
+		return err
+	}
+
+	mapper, typer := f.Object()
+	rmap := &resource.Mapper{
+		ObjectTyper:  typer,
+		RESTMapper:   mapper,
+		ClientMapper: f.ClientMapperForCommand(),
+	}
+
+	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(filenames...).
+		//SelectorParam(selector).
+		ResourceTypeOrNameArgs(true, args...).
+		Latest()
+	if err != nil {
+		return err
+	}
+
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	r := b.Flatten().Do()
+	infos, err := r.Infos()
+	if err != nil {
+		return err
+	}
+
+	defaultVersion := cmdutil.OutputVersion(cmd, clientConfig.Version)
+	results := editResults{}
+	for {
+		obj, err := resource.AsVersionedObject(infos, false, defaultVersion)
+		if err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+
+		// TODO: add an annotating YAML printer that can print inline comments on each field,
+		//   including descriptions or validation errors
+
+		// generate the file to edit
+		buf := &bytes.Buffer{}
+		if err := results.header.WriteTo(buf); err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+		if err := printer.PrintObj(obj, buf); err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+		original := buf.Bytes()
+
+		// launch the editor
+		edit := editor.NewDefaultEditor()
+		edited, file, err := edit.LaunchTempFile("oc-edit-", ext, buf)
+		if err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+
+		// cleanup any file from the previous pass
+		if len(results.file) > 0 {
+			os.Remove(results.file)
+		}
+
+		glog.V(4).Infof("User edited:\n%s", string(edited))
+		lines, err := hasLines(bytes.NewBuffer(edited))
+		if err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+		if bytes.Equal(original, edited) {
+			if len(results.edit) > 0 {
+				preservedFile(nil, file, cmd.Out())
+			} else {
+				os.Remove(file)
+			}
+			fmt.Fprintln(cmd.Out(), "Edit cancelled, no changes made.")
+			return nil
+		}
+		if !lines {
+			if len(results.edit) > 0 {
+				preservedFile(nil, file, cmd.Out())
+			} else {
+				os.Remove(file)
+			}
+			fmt.Fprintln(cmd.Out(), "Edit cancelled, saved file was empty.")
+			return nil
+		}
+
+		results = editResults{
+			file: file,
+		}
+
+		// parse the edited file
+		updates, err := rmap.InfoForData(edited, "edited-file")
+		if err != nil {
+			results.header.reasons = append(results.header.reasons, editReason{
+				head: fmt.Sprintf("The edited file had a syntax error: %v", err),
+			})
+			continue
+		}
+
+		visitor := resource.NewFlattenListVisitor(updates, rmap)
+
+		// need to make sure the original namespace wasn't changed while editing
+		if err = visitor.Visit(resource.RequireNamespace(cmdNamespace)); err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+
+		// attempt to calculate a delta for merging conflicts
+		delta, err := jsonmerge.NewDelta(original, edited)
```

Comment:
Specifically: https://github.com/GoogleCloudPlatform/kubernetes/blob/ed985b86544f4b129c724d160f9c300e9203eb1a/docs/api-conventions.md#strategic-merge-patch

This is what we use for `kubectl patch` and is what we plan to use for `kubectl apply`.

The key difference is that maps encoded as lists are merged like maps, by merging elements with matching keys.